File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        int n = spells.length;
        int m = potions.length;
        int[] ans = new int[n];

        for (int i = 0; i < n; i++) {
            int left = 0;
            int right = m - 1;
            while (left <= right) {
                int mid = (left + right) / 2;
                long product = (long) spells[i] * potions[mid];
                if (product >= success) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            ans[i] = m - left;
        }
        return ans;
    }
}

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(
        self, spells: List[int], potions: List[int], success: int
    ) -> List[int]:
        potions.sort()

        n = len(spells)
        m = len(potions)

        pairs = [0] * n

        for i in range(n):
            left = 0
            right = m - 1
            temp = 0

            while left <= right:
                mid = left + (right - left) // 2
                if spells[i] * potions[mid] >= success:
                    temp = m - mid
                    right = mid - 1
                else:
                    left = mid + 1

            pairs[i] = temp

        return pairs
